framework/math/ParallelVector/ParallelVector.h

```diff
@@ -0,0 +1,164 @@
+  virtual std::unique_ptr<ParallelVector> MakeNewVector() const = 0;
+
+  /** Returns the raw stl-vector associated with the local data of this
+   * vector.*/
+  virtual std::vector<double>& RawValues() = 0;
+
+  /// Return the size of the locally owned portion of the parallel vector.
+  uint64_t LocalSize() const { return local_size_; }
+
+  /// Return the global size of the parallel vector.
+  uint64_t GlobalSize() const { return global_size_; }
+
+  /**
+   * Read only accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double operator[](int64_t local_id) const = 0;
+
+  /**
+   * Read/write accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double& operator[](int64_t local_id) = 0;
+
+  /// Return a vector containing the locally owned data.
+  virtual std::vector<double> MakeLocalVector() = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given value.
+   */
+  virtual void Set(double value) = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given STL vector.
+   */
+  virtual void Set(const std::vector<double>& local_vector) = 0;
+
+  /**
+   * Define a set or add operation for the given global id-value pair
+   *
+   * This routine adds the global id-value pair to the set operation cache,
+   * which upon execution of Assemble, communicates the operations to the
+   * appropriate process.
+   */
+  virtual void SetValue(int64_t global_id, double value, VecOpType op_type) = 0;
+
+  /**
+   * Group multiple operations into a single call.
+   *
+   * This routine goes through the given global id-value pairs and calls
+   * SetValue for each.
+   */
+  virtual void SetValues(const std::vector<int64_t>& global_ids,
+                         const std::vector<double>& values,
+                         VecOpType op_type) = 0;
+
+  /**In place adding of vectors. The sizes must be compatible.*/
+  virtual void operator+=(const ParallelVector& y) = 0;
+
+  /**Adds a vector multiplied by scalar a. Optimized for a=1.0 and -1.0*/
+  virtual void PlusAY(double a, const ParallelVector& y) = 0;
+
+  /**Sets the local values of one vector equal to another. The sizes must be
+   * compatible.*/
+  virtual void CopyValues(const ParallelVector& y) = 0;
+
+  /**Copies a contiguous block of local data from the source vector to the
+   * current vector. The blocks are specified with offset values and num_values.
+   * */
+  virtual void BlockCopyLocalValues(const ParallelVector& y,
```

Comment:
`Set` overload.

Reply:
Losing a lot of short-hand info. If we have that many overloaded `Set`, you can basically get very little from auto-complete.

Reply:
Once you have `x.Set(` all of the different overloads show up in CLion and I'm sure in VSCode. This is something we've all seen when calling things like `x.resize(` on an STL vector in your favorite IDE.

Reply:
I am sorry I just don't agree. BlockSet here is a way better option and pops out with the autocomplete as "Yes, thats what I want".

Reply:
Okay, that's fine